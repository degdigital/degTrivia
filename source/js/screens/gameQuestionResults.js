import {replaceContent} from '../utils/domUtils.js';

function renderChoiceResults(results, correctChoice) {
	const chosenPercentages = calculateChosenPercentages(results);
	return Object.keys(results).reduce((html, optId) => {
		const resultCssClasses = ['choice-result'];
		
		if(optId.toString() === correctChoice.toString()) {
			resultCssClasses.push('choice-result--is-correct');
		}

		const resultHtml = `
			<div class="${resultCssClasses.join(' ')}">
				<span>${results[optId].text}</span>
				<span>${chosenPercentages[optId]}</span>
			</div>
		`;
	
		return html + resultHtml;
	}, '');
}

function calculateChosenPercentages(results) {
	const resultsKeys = Object.keys(results);
	const total = resultsKeys.reduce((output, key) => output + results[key].chosenCount, 0);
	let output = {};
	resultsKeys.forEach(key => {
		output[key] = Math.round((results[key].chosenCount / total) * 100) + '%';
	});
	return output;
}

function renderScreen(element, data) {
	const {questionData} = data;

	const html = `
		<div>
			<h1>Question #${questionData.order + 1}</h1>			
			<p>${questionData.question}</p>
			<div class="choice-results">
				${renderChoiceResults(questionData.choices, questionData.correctChoice)}
			</div>
		</div>`;

	replaceContent(element, html);
}

export default function({element}) {

	return {
		render: (data) => renderScreen(element, data)
	};
}