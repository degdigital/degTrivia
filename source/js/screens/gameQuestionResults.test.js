import gameQuestionResults from './gameQuestionResults';

let element;

const data = {
	questionData: {
		id: 10,
		order: 0,
		question: 'Who is the most dashing Kansas City Royal of all time?',
		choiceResults: [
			{
				choiceText: 'Bob Hamelin',
				isCorrectChoice: false,
				chosenCount: 28
			},
			{
				choiceText: 'Danny Tartabul',
				isCorrectChoice: false,
				chosenCount: 39
			},
			{
				choiceText: 'Steve Balboni',
				isCorrectChoice: true,
				chosenCount: 142
			}
		]
	}
};

beforeEach(() => {
	element = document.createElement('div');
	document.body.appendChild(element);
});

describe('calling render()', () => {
	test('renders out a question and choice results', async () => {
		await gameQuestionResults({element}).render(data);
		expect(element).toMatchSnapshot();
	});
});