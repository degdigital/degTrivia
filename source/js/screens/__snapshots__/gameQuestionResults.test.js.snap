// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render() renders out a question and choice results 1`] = `
<div>
  
		
  <div>
    
			
    <h1>
      Question #1
    </h1>
    			
			
    <p>
      Who is the most dashing Kansas City Royal of all time?
    </p>
    
			
			
    <div
      class="choice-results"
    >
      
				
			
      <div
        class="choice-result"
      >
        
				
        <span>
          Bob Hamelin
        </span>
        
				
        <span>
          13%
        </span>
        
			
      </div>
      
		
			
      <div
        class="choice-result"
      >
        
				
        <span>
          Danny Tartabul
        </span>
        
				
        <span>
          19%
        </span>
        
			
      </div>
      
		
			
      <div
        class="choice-result choice-result--is-correct"
      >
        
				
        <span>
          Steve Balboni
        </span>
        
				
        <span>
          68%
        </span>
        
			
      </div>
      
		
			
    </div>
    
		
  </div>
</div>
`;
